File: accsaber-api/src/main/java/de/ixsen/accsaber/api/controllers/RankedMapsController.java
package de.ixsen.accsaber.api.controllers;

import com.fasterxml.jackson.core.JsonProcessingException;
import de.ixsen.accsaber.api.dtos.CreateRankedMapDto;
import de.ixsen.accsaber.api.dtos.RankedMapDto;
import de.ixsen.accsaber.api.dtos.RankedMapsStatisticsDto;
import de.ixsen.accsaber.api.mapping.MappingComponent;
import de.ixsen.accsaber.business.PlayerService;
import de.ixsen.accsaber.business.RankedMapService;
import de.ixsen.accsaber.database.model.maps.BeatMap;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/ranked-maps")
public class RankedMapsController {

    private final RankedMapService rankedMapService;
    private final PlayerService playerService;
    private final MappingComponent mappingComponent;

    @Autowired
    public RankedMapsController(RankedMapService rankedMapService, PlayerService playerService, MappingComponent mappingComponent) {
        this.rankedMapService = rankedMapService;
        this.playerService = playerService;

        this.mappingComponent = mappingComponent;
    }

    @PostMapping
    public ResponseEntity<?> addNewRankedMap(@RequestBody CreateRankedMapDto rankedMapDto) {
        this.rankedMapService.addNewRankedMap(rankedMapDto.getBeatSaverKey(), rankedMapDto.getLeaderboardId(), rankedMapDto.getDifficulty(), rankedMapDto.getComplexity(), rankedMapDto.getCategoryName());
        this.playerService.recalculateApForAllPlayers();
        return ResponseEntity.noContent().build();
    }

    @GetMapping
    public ResponseEntity<List<RankedMapDto>> getAllRankedMaps() {
        List<RankedMapDto> rankedMapDtos = this.mappingComponent
                .getRankedMapMapper().rankedMapsToDtos(this.rankedMapService.getRankedMaps());
        return ResponseEntity.ok(rankedMapDtos);
    }

    @DeleteMapping("/{leaderboardId}")
    public ResponseEntity<?> removeRankedMap(@PathVariable Long leaderboardId) {
        this.rankedMapService.removeRankedMap(leaderboardId);
        this.playerService.recalculateApForAllPlayers();
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/statistics")
    public ResponseEntity<RankedMapsStatisticsDto> getRankedStat() {
        List<BeatMap> beatMaps = this.rankedMapService.getRankedMaps();
        long trueAccCount = beatMaps.stream().filter(r -> r.getComplexity() <= 4).count();
        long techAccCount = beatMaps.stream().filter(r -> r.getComplexity() >= 10).count();
        RankedMapsStatisticsDto rankedMapsStatisticsDto = new RankedMapsStatisticsDto();
        rankedMapsStatisticsDto.setMapCount(beatMaps.size());
        rankedMapsStatisticsDto.setTechAccMapCount(techAccCount);
        rankedMapsStatisticsDto.setTrueAccMapCount(trueAccCount);
        rankedMapsStatisticsDto.setStandardAccMapCount(beatMaps.size() - trueAccCount - techAccCount);
        Map<Double, Long> complexityToMapCount = beatMaps
                .stream()
                .collect(Collectors.groupingBy(BeatMap::getComplexity, Collectors.counting()));
        rankedMapsStatisticsDto.setComplexityToMapCount(complexityToMapCount);

        return ResponseEntity.ok(rankedMapsStatisticsDto);
    }

    @GetMapping("/{leaderboardId}")
    public ResponseEntity<RankedMapDto> getRankedMapDetails(@PathVariable Long leaderboardId) {
        BeatMap beatMap = this.rankedMapService.getRankedMap(leaderboardId);
        RankedMapDto rankedMapDto = this.mappingComponent.getRankedMapMapper().rankedMapToDto(beatMap);
        return ResponseEntity.ok(rankedMapDto);
    }

    @GetMapping("/playlist")
    public ResponseEntity<byte[]> getRankedMapPlaylist() throws JsonProcessingException {
        byte[] playlistJson = this.rankedMapService.getRankedMapsJson();
        return ResponseEntity
                .ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=accsaber-rankedmaps.json")
                .body(playlistJson);
    }
}

File: accsaber-api/src/main/kotlin/de/ixsen/accsaber/api/controllers/RankedMapsController.kt
package de.ixsen.accsaber.api.controllers

import com.fasterxml.jackson.core.JsonProcessingException
import de.ixsen.accsaber.api.dtos.CreateRankedMapDto
import de.ixsen.accsaber.api.dtos.RankedMapDto
import de.ixsen.accsaber.api.dtos.RankedMapsStatisticsDto
import de.ixsen.accsaber.api.dtos.UpdateComplexityDto
import de.ixsen.accsaber.api.mapping.RankedMapMapper
import de.ixsen.accsaber.business.CategoryService
import de.ixsen.accsaber.business.PlayerService
import de.ixsen.accsaber.business.PlaylistService
import de.ixsen.accsaber.business.RankedMapService
import de.ixsen.accsaber.database.model.maps.BeatMap
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/ranked-maps")
class RankedMapsController @Autowired constructor(
    private val rankedMapService: RankedMapService,
    private val categoryService: CategoryService,
    private val playerService: PlayerService,
    private val playlistService: PlaylistService,
    private val rankedMapMapper: RankedMapMapper,
) {
    // FIXME The recognition should actually try to evaluate if the hash is a valid hash
    @PostMapping
    fun addNewRankedMapByKey(@RequestBody rankedMapDto: CreateRankedMapDto): ResponseEntity<*> {
        if (rankedMapDto.id!!.length > 10) {
            this.rankedMapService.addNewRankedMapByHash(rankedMapDto.id, rankedMapDto.difficulty!!, rankedMapDto.complexity!!, rankedMapDto.categoryName!!)
        } else {
            this.rankedMapService.addNewRankedMapByKey(rankedMapDto.id, rankedMapDto.difficulty!!, rankedMapDto.complexity!!, rankedMapDto.categoryName!!)
        }
        return ResponseEntity.noContent().build<Any>()
    }

    @PutMapping("{leaderboardId}")
    fun updateComplexity(@RequestBody updateComplexityDto: UpdateComplexityDto, @PathVariable leaderboardId: String) {
        this.rankedMapService.updateComplexity(updateComplexityDto.leaderboardId, updateComplexityDto.complexity)
    }

    @PutMapping
    fun updateComplexities(@RequestBody updateComplexityDtos: List<UpdateComplexityDto>) {
        updateComplexityDtos.forEach { this.rankedMapService.updateComplexity(it.leaderboardId, it.complexity) }
    }

    @GetMapping
    fun getAllRankedMaps(): ResponseEntity<List<RankedMapDto>> {
        val rankedMapDtos = this
            .rankedMapMapper.rankedMapsToDtos(this.rankedMapService.getAllRankedMaps())
        return ResponseEntity.ok(rankedMapDtos)
    }

    @GetMapping("/category/{categoryName}")
    fun getAllRankedMapsForCategory(@PathVariable categoryName: String): ResponseEntity<List<RankedMapDto>> {
        val category = categoryService.getCategoryByName(categoryName)
        val categoryRankedMaps = this.rankedMapService.getAllRankedMapsForCategory(category)
        val rankedMapDtos = this.rankedMapMapper.rankedMapsToDtos(categoryRankedMaps)
        return ResponseEntity.ok(rankedMapDtos)
    }

    @DeleteMapping("/{leaderboardId}")
    fun removeRankedMap(@PathVariable leaderboardId: Long): ResponseEntity<*> {
        this.rankedMapService.removeRankedMap(leaderboardId)
        playerService.recalculateApForAllPlayers()
        return ResponseEntity.noContent().build<Any>()
    }

    @GetMapping("/statistics")
    fun getRnkedStat(): ResponseEntity<RankedMapsStatisticsDto> {
        val beatMaps = this.rankedMapService.getAllRankedMaps()
        val trueAccCount = beatMaps.stream().filter { r: BeatMap -> r.complexity <= 4 }.count()
        val techAccCount = beatMaps.stream().filter { r: BeatMap -> r.complexity >= 10 }.count()

        val complexityToMapCount = beatMaps.groupingBy { it.complexity }.eachCount()

        val rankedMapsStatisticsDto = RankedMapsStatisticsDto(
            beatMaps.size, trueAccCount, beatMaps.size - trueAccCount - techAccCount, techAccCount, complexityToMapCount
        )

        return ResponseEntity.ok(rankedMapsStatisticsDto)
    }

    @GetMapping("/{leaderboardId}")
    fun getRankedMapDetails(@PathVariable leaderboardId: Long): ResponseEntity<RankedMapDto> {
        val beatMap = this.rankedMapService.getRankedMap(leaderboardId)
        val rankedMapDto = this.rankedMapMapper.rankedMapToDto(beatMap)
        return ResponseEntity.ok(rankedMapDto)
    }

    @Throws(JsonProcessingException::class)
    @GetMapping("/playlist")
    fun rankedMapPlaylist(): ResponseEntity<ByteArray> {
        val playlistJson = playlistService.getPlaylist("all")
        return ResponseEntity
            .ok()
            .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=accsaber-rankedmaps.json")
            .body(playlistJson)
    }
}